packages/gatsby/src/schema/infer/infer.js

```diff
@@ -0,0 +1,301 @@
+const _ = require(`lodash`)
+const {
+  GraphQLList,
+  GraphQLObjectType,
+  defaultFieldResolver,
+  getNamedType,
+} = require(`graphql`)
+const invariant = require(`invariant`)
+
+const { isFile } = require(`./is-file`)
+const { link, fileByPath } = require(`../resolvers`)
+const { isDate, dateResolver } = require(`../types/Date`)
+const is32BitInteger = require(`./is-32-bit-integer`)
+
+const { store } = require(`../../redux`)
+
+const addInferredFields = ({
+  schemaComposer,
+  typeComposer,
+  exampleValue,
+  nodeStore,
+  parentSpan,
+}) =>
+  addInferredFieldsImpl({
+    schemaComposer,
+    typeComposer,
+    nodeStore,
+    exampleObject: exampleValue,
+    prefix: typeComposer.getTypeName(),
+    depth: 0,
+  })
+
+module.exports = {
+  addInferredFields,
+}
+
+const addInferredFieldsImpl = ({
+  schemaComposer,
+  typeComposer,
+  nodeStore,
+  exampleObject,
+  prefix,
+  depth,
+}) => {
+  const fields = {}
+  Object.keys(exampleObject).forEach(unsanitizedKey => {
+    const exampleValue = exampleObject[unsanitizedKey]
+    let key = createFieldName(unsanitizedKey)
+    const selector = `${prefix}.${key}`
+
+    let arrays = 0
+    let value = exampleValue
+    while (Array.isArray(value)) {
+      value = value[0]
+      arrays++
+    }
+
+    let fieldConfig
+    if (hasMapping(selector)) {
+      // TODO: Use `prefix` instead of `selector` in hasMapping and getFromMapping?
+      // i.e. does the config contain sanitized field names?
+      fieldConfig = getFieldConfigFromMapping(selector)
+    } else if (key.includes(`___NODE`)) {
+      fieldConfig = getFieldConfigFromFieldNameConvention(
+        schemaComposer,
+        nodeStore,
+        exampleValue,
+        unsanitizedKey
+      )
+      key = key.split(`___NODE`)[0]
+    } else {
+      fieldConfig = getFieldConfig(
+        schemaComposer,
+        nodeStore,
+        value,
+        selector,
+        depth
+      )
+    }
+
+    while (arrays--) {
+      fieldConfig = { ...fieldConfig, type: [fieldConfig.type] }
+    }
+
+    // Proxy resolver to unsanitized fieldName in case it contained invalid characters
+    if (key !== unsanitizedKey) {
+      // Don't create a field with the sanitized key if a field with that name already exists
+      if (!(exampleObject[key] == null && !typeComposer.hasField(key))) {
```

Comment:
Should we keep the `invariant` here? I.e. for the case when the sanitized field name already exists on the type.

Reply:
I've removed invariant because I think some behavior in www expected it to just be ignored, but I'm not 100% sure now.

Reply:
Might have been #11578

Reply:
I'll try.